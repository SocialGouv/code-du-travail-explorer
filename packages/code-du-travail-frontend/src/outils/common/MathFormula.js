import React from "react";
import PropTypes from "prop-types";
import { BlockMath } from "react-katex";
import { theme } from "@socialgouv/react-ui";
import styled from "styled-components";

export const asciiMathToTex = (ascii) => {
  // multiplications
  let tex = ascii.replace(/\*/g, `\\times`);
  // Divisions
  /* eslint-disable no-useless-escape */
  const parenthesisMatcher = `(\\(.+\\))`;
  const nonParenthesisMatcher = `([^()\\s]+)`;
  const fullMatcher = `(?:${parenthesisMatcher}|${nonParenthesisMatcher})`;
  while (tex.includes("/")) {
    tex = tex.replace(
      new RegExp(`${fullMatcher}\\s?\\/\\s?${fullMatcher}`, "gm"),
      `\\frac{$1$2}{$3$4}`
    );
  }
  /* eslint-enable */
  return tex;
};

export const MathFormula = ({ formula }) => {
  return (
    <Wrapper>
      <StyledFormula>
        <BlockMath>{asciiMathToTex(formula)}</BlockMath>
      </StyledFormula>
    </Wrapper>
  );
};

MathFormula.propTypes = {
  formula: PropTypes.string.isRequired,
};

const { spacings } = theme;

const Wrapper = styled.div`
  margin: ${spacings.base} 0;
`;

const StyledFormula = styled.div`
  display: flex;
`;

export default MathFormula;
