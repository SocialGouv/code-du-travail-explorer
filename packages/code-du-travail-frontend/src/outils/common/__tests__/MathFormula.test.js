import React from "react";
import { render } from "@testing-library/react";
import { asciiMathToTex, MathFormula } from "../MathFormula";

const formulae = `26 + (this (is) weird) / (45 (why not)) + 3

 46.7 / 65.2 * 6

47 * (this should be)/okok

(1 / 3) /4 * 7 * 8

 4 / (1 / 3) * 10

6 / (sref)

6 / 3

 no no / ok ok ok `;

describe("asciiMathToTex", () => {
  it("gives the correct TeX formula provided a correct ascii input", () => {
    expect(asciiMathToTex(formulae)).toMatchSnapshot();
  });
});
describe("<MathFormula />", () => {
  it("renders", () => {
    const { container } = render(
      <MathFormula formula="1 / 4 * Sref * A * 2"></MathFormula>
    );
    expect(container).toMatchSnapshot();
  });
});
