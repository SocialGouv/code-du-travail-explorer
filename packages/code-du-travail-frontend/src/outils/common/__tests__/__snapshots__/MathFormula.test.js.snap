// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MathFormula /> renders 1`] = `
.c0 {
  margin: 1.6rem 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div>
        <span
          class="katex-display"
        >
          <span
            class="katex"
          >
            <span
              class="katex-mathml"
            >
              <math>
                <semantics>
                  <mrow>
                    <mfrac>
                      <mrow>
                        <mn>
                          1
                        </mn>
                      </mrow>
                      <mrow>
                        <mn>
                          4
                        </mn>
                      </mrow>
                    </mfrac>
                    <mo>
                      ×
                    </mo>
                    <mi>
                      S
                    </mi>
                    <mi>
                      r
                    </mi>
                    <mi>
                      e
                    </mi>
                    <mi>
                      f
                    </mi>
                    <mo>
                      ×
                    </mo>
                    <mi>
                      A
                    </mi>
                    <mo>
                      ×
                    </mo>
                    <mn>
                      2
                    </mn>
                  </mrow>
                  <annotation
                    encoding="application/x-tex"
                  >
                    \\frac{1}{4} \\times Sref \\times A \\times 2
                  </annotation>
                </semantics>
              </math>
            </span>
            <span
              aria-hidden="true"
              class="katex-html"
            >
              <span
                class="strut"
                style="height:1.32144em;"
              />
              <span
                class="strut bottom"
                style="height:2.00744em;vertical-align:-0.686em;"
              />
              <span
                class="base"
              >
                <span
                  class="mord"
                >
                  <span
                    class="mopen nulldelimiter"
                  />
                  <span
                    class="mfrac"
                  >
                    <span
                      class="vlist-t vlist-t2"
                    >
                      <span
                        class="vlist-r"
                      >
                        <span
                          class="vlist"
                          style="height:1.32144em;"
                        >
                          <span
                            style="top:-2.314em;"
                          >
                            <span
                              class="pstrut"
                              style="height:3em;"
                            />
                            <span
                              class="mord"
                            >
                              <span
                                class="mord"
                              >
                                4
                              </span>
                            </span>
                          </span>
                          <span
                            style="top:-3.15em;"
                          >
                            <span
                              class="pstrut"
                              style="height:3em;"
                            />
                            <span
                              class="stretchy"
                              style="height:0.2em;"
                            >
                              <svg
                                height="0.2em"
                                preserveAspectRatio="xMinYMin slice"
                                viewBox="0 0 400000 200"
                                width="400em"
                              >
                                <path
                                  d="M0 80H400000 v40H0z M0 80H400000 v40H0z"
                                />
                              </svg>
                            </span>
                          </span>
                          <span
                            style="top:-3.677em;"
                          >
                            <span
                              class="pstrut"
                              style="height:3em;"
                            />
                            <span
                              class="mord"
                            >
                              <span
                                class="mord"
                              >
                                1
                              </span>
                            </span>
                          </span>
                        </span>
                        <span
                          class="vlist-s"
                        >
                          ​
                        </span>
                      </span>
                      <span
                        class="vlist-r"
                      >
                        <span
                          class="vlist"
                          style="height:0.686em;"
                        />
                      </span>
                    </span>
                  </span>
                  <span
                    class="mclose nulldelimiter"
                  />
                </span>
                <span
                  class="mord rule"
                  style="margin-right:0.2222222222222222em;"
                />
                <span
                  class="mbin"
                >
                  ×
                </span>
                <span
                  class="mord rule"
                  style="margin-right:0.2222222222222222em;"
                />
                <span
                  class="mord mathit"
                  style="margin-right:0.05764em;"
                >
                  S
                </span>
                <span
                  class="mord mathit"
                  style="margin-right:0.02778em;"
                >
                  r
                </span>
                <span
                  class="mord mathit"
                >
                  e
                </span>
                <span
                  class="mord mathit"
                  style="margin-right:0.10764em;"
                >
                  f
                </span>
                <span
                  class="mord rule"
                  style="margin-right:0.2222222222222222em;"
                />
                <span
                  class="mbin"
                >
                  ×
                </span>
                <span
                  class="mord rule"
                  style="margin-right:0.2222222222222222em;"
                />
                <span
                  class="mord mathit"
                >
                  A
                </span>
                <span
                  class="mord rule"
                  style="margin-right:0.2222222222222222em;"
                />
                <span
                  class="mbin"
                >
                  ×
                </span>
                <span
                  class="mord rule"
                  style="margin-right:0.2222222222222222em;"
                />
                <span
                  class="mord"
                >
                  2
                </span>
              </span>
            </span>
          </span>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`asciiMathToTex gives the correct TeX formula provided a correct ascii input 1`] = `
"26 + \\\\frac{(this (is) weird)}{(45 (why not))} + 3

 \\\\frac{46.7}{65.2} \\\\times 6

47 \\\\times \\\\frac{(this should be)}{okok}

\\\\frac{(\\\\frac{1}{3})}{4} \\\\times 7 \\\\times 8

 \\\\frac{4}{(\\\\frac{1}{3})} \\\\times 10

\\\\frac{6}{(sref)}

\\\\frac{6}{3}

 no \\\\frac{no}{ok} ok ok "
`;
